### ISSUE TITLE
link remains harmful

### ISSUE THREAD

**Issue by opoudjis** [metanorma/metanorma-standoc#35]
This is something of a Heisenbug, hard to reproduce, but:

`https://www.iso.org/directives[www.iso.org/directives]` on the first iteration of processing links is being parsed properly into `<link target="https://www.iso.org/directives"> www.iso.org/directives</link>`. However for reasons I cannot reproduce or understand, on the second iteration of Nokogiri processing, when the XML is serialised, it is ending up as `<link target="https://www.iso.org/directives"/>www.iso.org/directives`. 

There is something, likely in Nokogiri, which is getting stuck on `link` and mis-serialising it. I have had previous such problems with the same element name (https://github.com/riboseinc/metanorma-iso/issues/187). I will have to put in a workaround of processing links in Nokogiri as a different element name (using `elink` makes the problem go away), and then convert it back to `link` on cleanup.

**Comment by opoudjis** [comment]
Have replicated, in reading in of XML file with `<link>`.

**Comment by opoudjis** [comment]
Example found by @manuel489 

[link-issues.zip](https://github.com/metanorma/metanorma-standoc/files/3799939/link-issues.zip)

**Comment by ronaldtse** [comment]
Hope this gets fixed soon. I have always had an issue explaining to people why links don’t work.

**Comment by opoudjis** [comment]
I spent 4 hours last time I tried to fix this, and it was random behaviour deep inside Asciidoctor that I could not explain. I have just upgraded to Asciidoctor 2.0.10; let see if that made a difference...

**Comment by opoudjis** [comment]
Now this is interesting, though it may just go to the Heisenbug nature of this bug:

* In ISO, Asciidoctor is converting the link:

````
link:http://www.tc184-sc4.org/titles/OTD_Titles.htm[URL label]
```` 

````
<p id="_aa9d8f3d-5f7f-46fb-a0c8-33bfa91939b3">URL sample: <link target="http://www.tc184-sc4.org/titles/OTD_Titles.htm"/>URL label</p>
````

This is our longstanding issue. In ITU, the same document converts as:

````
<p id="_b257f278-aca2-45b8-9a35-3276324fe215">URL sample: <link target="http://www.tc184-sc4.org/titles/OTD_Titles.htm">URL label</link></p>
````

**Comment by opoudjis** [comment]
The bug is in Nokogiri, in the multiple conversions backwards and forwards between Nokogiri API and XML during Nokogiri Builder. But that leaves me none the wiser yet...

**Comment by opoudjis** [comment]
As feared: if I stepped through the debugger more carefully, with no code change, the bug disappears.

**Comment by opoudjis** [comment]
I have found what seems to be the issue, though it now needs even more digging:

When Nokogiri adds nodes of XML to builder, sometimes the `<title>` of an XML element is escaped, by having `=` to the tag:

````
<foreword> \n  <title>Foreword</title> \n  <p id=\"_ed9db58b-cab3-4db1-bcd1-69c7e1b892f3\">URL sample: <link target=\"http://www.tc184-sc4.org/titles/OTD_Titles.htm\">URL label</link></p> \n</foreword>\n<introduction id=\"_introduction\"> \n  <title=>Introduction</title=> \n  <p id=\"_c5272acd-860c-40f2-9540-25cd9a7ec19f\">Random text</p> \n</introduction>
````

Note that I am not making introduction/title be `<title=>`, Nokogiri is.

When that block of XML is added into `<sections>`, the `<link>` element is mangled, as we have seen. If I correct the `<title=>` tag to `<title>`, it is not.

Whatever is inserting the = in there, it gets removed by the time the XML is output.

Rather than fix that behaviour in whatever obscure corner of Nokogiri it happens, I'll preempt it:

````
def introduction_parse(attrs, xml, node)
        xml.introduction **attr_code(attrs) do |xml_section|
          xml_section.title = "Introduction"
          content = node.content
          xml_section << content
        end
      end
````

Setting any element in Nokogiri with = instead of a block (`xml_section.title { |t| t << "Normative References" }`) is triggering this mangling of links. Presumably, by printing more details of the XML structure during debugging, I was triggering the cleanup Nokogiri must be doing to remove the `=` from tags.

Changing the equals assignment to the Introduction title seems to be removing this bug. I will clean up the stack, but I'm pretty sure this is one of the few places this occurs, since it is some of the earliest code I wrote.
